draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -3614,6 +3625,112 @@ Steps:
 38. return (xEn, xEd, yEn, yEd)
 ~~~
 
+## Cofactor Clearing for BLS12-381 G2 {#clear-cofactor-bls12381-g2}
+
+The curve BLS12-381, whose parameters are defined in {{suites-bls12381-g2}},
+admits an efficiently-computable endomorphism psi that can be used to
+speed up cofactor clearing for G2 {{SBCDK09}} {{FKR11}} {{BP17}} (see also
+{{cofactor-clearing}}).
+This section implements the endomorphism psi and a fast cofactor clearing
+method described by Budroni and Pintore {{BP17}}.
+
+The functions in this section operate on points whose coordinates are
+represented as ratios, i.e., (xn, xd, yn, yd) corresponds to the point
+(xn / xd, yn / yd); see {{projective-coords}} for further discussion of
+projective coordinates.
+When points are represented in affine coordinates, one can simply ignore
+the denominators (xd == 1 and yd == 1).
+
+The following function computes the Frobenius endomorphism for an element
+of F = GF(p^2) with basis (1, I), where I^2 + 1 == 0 in F.
+(This is the base field of the BLS12-381 curve containing the subgroup G2.)
+
+~~~
+frobenius(x)
+
+Input: x, an element of GF(p^2).
+Output: a, an element of GF(p^2).
+
+Notation: x = x0 + I * x1, where x0 and x1 are elements of GF(p).
+
+Steps:
+1. a = x0 - I * x1
+2. return a
+~~~
+
+The following function computes the endomorphism psi for points on the BLS12-381
+curve defined over GF(p^2).
+
+~~~
+psi(xn, xd, yn, yd)
+
+Input: P, the point (xn / xd, yn / yd) on the BLS12-381 curve
+       containing the subgroup G2.
+Output: Q, a point on the same curve.
+
+Constants:
+1. c1 = 1 / (1 + I)^((p - 1) / 3)           # in GF(p^2)
+2. c2 = 1 / (1 + I)^((p - 1) / 2)           # in GF(p^2)
+
+Steps:
+1. qxn = c1 * frobenius(xn)
+2. qxd = frobenius(xd)
+3. qyn = c2 * frobenius(yn)
+4. qyd = frobenius(yd)
+5. return (qxn, qxd, qyn, qyd)
+~~~
+
+The following function efficiently computes psi(psi(P)).
+
+~~~
+psi2(xn, xd, yn, yd)
+
+Input: P, the point (xn / xd, yn / yd) on the BLS12-381 curve
```

Comment:
I think we should specify that the point is on the twist of the BLS12-381, since the twist is a different curve.

Reply:
I think it's a great idea to make clear that this is a different curve. But the term "twist" isn't used anywhere in the document, and it seems like a shame to either use it undefined or need to add a definition here. It would be precise to say "the curve containing the subgroup G2"---would that work for you?

Reply:
The term doesn't need to be defined as long as the appropriate references to pairing draft are made.

>  "the curve containing the subgroup G2"

In fact, G2 lives in the original curve but over Fp12 (which also satisfies that sentence)

Reply:
> In fact, G2 lives in the original curve but over Fp12 (which also satisfies that sentence)

Hmm, that does not seem to be how G2 is defined anywhere. It's certainly true that this subgroup is isomorphic to a subgroup of the curve defined over Fp12, but in all usage I have seen (including in this draft) elements of the group G2 are points on a curve defined over Fp2.

As further support, Budroni and Pintore also use G2 to mean points on the curve defined over Fp2, as does prior work in this line.

It is better, in my opinion, to use a slightly longer description that is easily understood rather than a description that the reader may not grok without consulting other references.

Reply:
But actually I believe there's an easy way to avoid this issue entirely, just by talking directly about the curve we mean rather than some description of that curve. I'll push a proposed edit in a moment and we can discuss further.

Reply:
The formal definition of G2 is the r-torsion points in the original curve over Fp^k.
For implementation purposes, it is common to use a degree-d twist of the curve over Fp^(k/d).
You will be agree with all this.

Reply:
For sure, what you say is absolutely right with respect to the way the pairing operation is defined, etc.! It's only a question of notation what we call elements of G2. But I'm sure that readers of this document don't need to worry about this technicality; it is least confusing, in my mind, to stick with the common usage of G2.